### PR TITLE
Bradley/flexcpu/mem forward

### PR DESCRIPTION
**Pull request by animalvgamer**

### CODE REVIEW

File: src/cpu/flexcpu/stld_forwarder.cc [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+    });
+    // TODO consider holding onto the store for a little longer (e.g. until
+    //      memory response time), so more loads can be forwarded. Pretty sure
+    //      the other CPU model LSQs will hold onto items until response anyway
+    //      but we haven't implemented it yet only to keep the code simple for
+    //      now.
+    inst_ptr->addCommitCallback([this] { commitStore(); });
+}
+
+// anonymous namespace for internal linkage only.
```
Comment (powerjg) [inline src/cpu/flexcpu/stld_forwarder.cc]: This should be in the header as part of the class description.

File: src/cpu/flexcpu/stld_forwarder.hh [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+#include <list>
+#include <memory>
+#include <unordered_map>
+
+#include "base/types.hh"
+#include "cpu/flexcpu/inflight_inst.hh"
+
+class StLdForwarder {
+  protected:
```
Comment (powerjg) [inline src/cpu/flexcpu/stld_forwarder.hh]: Planning on having this be a parent class?

File: src/cpu/flexcpu/stld_forwarder.hh [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+     *                 stack (before requestLoad() returns) if the latency for
+     *                 forwarding is configured to be zero. Packet is nullptr
+     *                 if forwarding is not possible.
+     */
+    void requestLoad(const std::shared_ptr<InflightInst>& inst_ptr,
+                     const RequestPtr& req,
+                     const std::function<void(PacketPtr)>& callback);
+};
+
+#endif // __CPU_FLEXCPU_LSQ_HH__
```
Comment (powerjg) [inline src/cpu/flexcpu/stld_forwarder.hh]: Probably should have a newline at the end of the file. Most editors are configured this way...

File: src/cpu/flexcpu/inflight_inst.cc [hunk 4]
```diff
@@ -215,6 +215,12 @@ InflightInst::addReadyCallback(function<void()> callback)
     readyCallbacks.push_back(callback);
 }
 
+void
+InflightInst::addRetireCallback(function<void()> callback)
```
[the remark this answers is not in the document]
Comment (animalvgamer) [reply src/cpu/flexcpu/inflight_inst.cc]: Oh that's actually a good call I didn't think of. I'll try it tomorrow.

File: src/cpu/flexcpu/stld_forwarder.cc (hunk 1 above)
Comment (animalvgamer) [reply src/cpu/flexcpu/stld_forwarder.cc]: Why? It's not important to the interface or internal class definition, so I thought it was a good idea to hide it with the function implementation.

File: src/cpu/flexcpu/stld_forwarder.hh (hunk 2 above)
Comment (animalvgamer) [reply src/cpu/flexcpu/stld_forwarder.hh]: I kind of default to protected over private? I did this to every other class in flexcpu/.

File: src/cpu/flexcpu/stld_forwarder.hh (hunk 3 above)
Comment (animalvgamer) [reply src/cpu/flexcpu/stld_forwarder.hh]: will change when I get in

File: src/cpu/flexcpu/inflight_inst.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (animalvgamer) [reply src/cpu/flexcpu/inflight_inst.cc]: I'll make this change in a separate commit/PR, so I can test variants for performance impact.

File: src/cpu/flexcpu/stld_forwarder.cc (hunk 1 above)
Comment (animalvgamer) [reply src/cpu/flexcpu/stld_forwarder.cc]: I moved the declaration into the only function where it's being used, so it's perhaps less weird?

**Pull request merged**